`Programming exercises`: Display checkout directories in build plan preview

florian-glombik opened the pull request:
### Checklist
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).


#### Server
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.


#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.


### Motivation and Context
With the introduction of localCI it is now opaque where the repositories are actually checked out. This information was previously available when viewing the respective build plan in Bamboo (or Jenkins).

The information can currently only be found by reading the documentation and referring to the actually used code.


### Description
#### Client
* Added checkout directory information to the programming exercise detail page
* Added checkout directory information to the programming exercise create/edit page _(in the preview section)_
* Removed method call from html in `programming-exercise-plans-and-repositories-preview.component.ts`
* Added undefined checks in `programming-exercise-update.component.ts` since console errors were displayed in the console when adding auxiliary repositories
* Introduced shared constant instead of duplicated RegEx
* Moved `build-log-statistics-dto.ts` to the `entities` folder, as we seem to store all dtos there in the client side
* Introduced new shared Module and Components for the repository + checkout directory table
* Removed method calls from the re-used repository + checkout directory table code

#### Server
* Added missing checkout directory information for the solution repository of some programming languages _(changes in `ContinousIntegrationService`)_
* Added endpoint to get the checkout directories for a programming exercise in a certain programming language


### Steps for Testing
The change will only be visible on a localCI test server!

Prerequisites:
- 1 Instructor

1. Create a new programming exercise
2. Define the exercise title and short name
3. See that the checkout directories will be displayed in the preview
4. Change the programming language and see that the checkout directories will be updated accordingly _(e.g. `Java` -> `OCaml`)_
5. Add an auxiliary repository (e.g. name: `sut`, checkoutDirectory: `assignment/sut`)
6. See that the checkout directory will be displayed aswell _(it is intended to add a `/` for the root directory for consistency if e.g. the `test` repository is checked out in the root repository)_
7. Remove the checkmark for `Check out repository of sample solution` (only available for OCaml and Haskell)
8. See that the solution repository is not checked out during template build
9. Save the exercise
10. See that the checkout directories are displayed in the details view of the exercise

If you additionally want to verify that the displayed information is checked you can set a breakpoint in `BuildJobContainerService.populateBuildJobContainer` and then look at the docker if container is build as expected from the information displayed in the new UI.
You will find the testing directory under `/var/tmp/testing-dir/` in the docker container.
_(Note: some build scripts only have images for amd, which will early abort as the image will be pulled first - either define an existing image for the exercise or change you application-local.yml accordingly, e.g. `continuous-integration.image-architecture: amd64`)_

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)

### Review Progress
#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [programming-exercise-detail.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6578/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/programming-exercise-detail.component.ts.html) | 85.6% | ✅ |
| [repositories-checkout-directories-dto.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6578/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/repositories-checkout-directories-dto.ts.html) | 0% | ❌ (simple DTO without further logic) |
| [programming-exercise.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6578/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/services/programming-exercise.service.ts.html) | 88.02% | ✅ |
| [programming-exercise-plans-and-repositories-preview.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6578/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/programming-exercise-plans-and-repositories-preview.component.ts.html) | 96.15% | ✅ |
| [utils.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6578/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/util/utils.ts.html) | 95.52% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ContinuousIntegrationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6578/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.ci/ContinuousIntegrationService.html) | 100% | ✅ |
| [LocalCIService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6578/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localci/LocalCIService.html) | 98% | ✅ |
| [RepositoriesCheckoutDirectoryDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6578/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto/RepositoriesCheckoutDirectoryDTO.html) | 100% | ❌ (simple DTO without further logic) |
| [ProgrammingExerciseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6578/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.programming/ProgrammingExerciseResource.html) | 81% | ✅ |


### Screenshots
#### Change Comparison
| Change                                     | Before                                                                                             | Now                                                                                                |
|--------------------------------------------|----------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------|
| Added checkout directories to preview      | ![image](https://github.com/ls1intum/Artemis/assets/63976129/49d16446-6aff-4953-bff3-d1e4618d08e4) | ![image](https://github.com/ls1intum/Artemis/assets/63976129/4bd7d0d6-ee05-4594-aaeb-a1de36e92346) |
| Added checkout directories to details view | ![image](https://github.com/ls1intum/Artemis/assets/63976129/69484603-de45-455d-9f8e-09aa9c7f2787) | ![image](https://github.com/ls1intum/Artemis/assets/63976129/a4d77c44-2700-47e0-b132-d3a8d11f82e0) |
| Adjusted tooltip for LocalCI setup         | ![image](https://github.com/ls1intum/Artemis/assets/63976129/1a7c3b59-ff47-4099-be66-c1038f2edef1) | ![image](https://github.com/ls1intum/Artemis/assets/63976129/5dc25ba5-6332-478f-94c3-44aa80f78c2a) |
| Further Tooltips                           |                                                                                                    | ![image](https://github.com/ls1intum/Artemis/assets/63976129/b14bfee9-ee7b-4127-a626-6f536135a174) ![image](https://github.com/ls1intum/Artemis/assets/63976129/dd6c1504-89b6-4e5c-ae31-495f11c580c8) |

#### Screencast
![checkoutDirectoryDemo](https://github.com/ls1intum/Artemis/assets/63976129/77ca54d2-80eb-4cf3-b3b3-b81dacc587e0)


_The screencast may take some time to load_ 🐌

florian-glombik commented on src/main/webapp/app/exercises/programming/manage/update/programming-exercise-plans-and-repositories-preview.component.html:
```diff
@@ -8,21 +8,15 @@
-                        <code class="text-white badge bg-primary label-preview">
-                            {{ getCourseShortName()?.toLowerCase() }}{{ programmingExercise.shortName.toLowerCase() }}-exercise
-                        </code>
+                        <code class="text-white badge bg-primary label-preview"> {{ shortName?.toLowerCase() }}{{ programmingExercise.shortName.toLowerCase() }}-exercise </code>
```
Using the variable `shortName` instead of the method `getCourseShortName` as the method will be executed on each change (bad performance!)

florian-glombik commented on src/main/webapp/app/exam/manage/exams/exam-exercise-import/exam-exercise-import.component.ts:
```diff
@@ -33,8 +33,8 @@ export class ExamExerciseImportComponent implements OnInit {
-    readonly shortNamePattern = RegExp('(^(?![\\s\\S]))|^[a-zA-Z][a-zA-Z0-9]*$|' + SHORT_NAME_PATTERN); // must start with a letter and cannot contain special characters
-    readonly titleNamePattern = RegExp('^[a-zA-Z0-9-_ ]+'); // must only contain alphanumeric characters, or whitespaces, or '_' or '-'
+    readonly SHORT_NAME_REGEX = RegExp('(^(?![\\s\\S]))|^[a-zA-Z][a-zA-Z0-9]*$|' + SHORT_NAME_PATTERN); // must start with a letter and cannot contain special characters
+    readonly TITLE_NAME_REGEX = RegExp(TITLE_NAME_PATTERN);
```
There is a console error with the current RegEx, while trying to fix it I have noticed that this RegEx was duplicated, so I introduced a constant for it.
I reverted the change to the RegEx since some tests failed, the adjustment I had in mind was not semantically equal. 

I left the changes since now the RegEx at least only needs to be fixed in one place.

florian-glombik commented on src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:
```diff
@@ -171,76 +183,105 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
-                                this.programmingExercise.projectKey,
-                                this.programmingExercise.solutionParticipation.buildPlanId,
-                            );
+            this.templateAndSolutionParticipationSubscription = this.programmingExerciseService
```
Only added `this. templateAndSolutionParticipationSubscription = [...]` here, due to the wrap to the next line there is more marked as diff than actually changed

florian-glombik commented on src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:
```diff
@@ -171,76 +183,105 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
+                        this.exerciseDetailSections = this.getExerciseDetails();
+                    });
+
+                    this.submissionPolicySubscription = this.programmingExerciseSubmissionPolicyService
```
Assigned another subscription here in order to unsubscribe in `ngOnDestroy`

florian-glombik commented on src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:
```diff
@@ -171,76 +183,105 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
-                this.exerciseDetailSections = this.getExerciseDetails();
-            });
+            if (this.programmingExercise.programmingLanguage) {
+                this.checkoutDirectoriesSubscription = this.programmingExerciseService
```
Added `this.checkoutDirectoriesSubscription = `

florian-glombik commented on src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:
```diff
@@ -343,6 +384,7 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
                         exerciseId: exercise.id,
                         type: 'TEMPLATE',
                         showOpenLink: !this.localVCEnabled,
+                        checkoutDirectory: this.localVCEnabled ? this.checkoutDirectories?.exerciseCheckoutDirectory : undefined,
```
If we are not on LocalCI we display the build plans instead (there the checkout directories can be looked up)

florian-glombik commented on src/main/webapp/app/exercises/programming/manage/update/programming-exercise-plans-and-repositories-preview.component.ts:
```diff
@@ -10,11 +15,69 @@ import { ProgrammingExercise } from 'app/entities/programming-exercise.model';
+        this.shortName = getCourseFromExercise(this.programmingExercise)?.shortName;
+    }
+
+    ngOnInit() {
```
As far as I understand this check complains about the import `import { OnInit } from '@angular/core';` not being present.

IntelliJ will display it as unused method and complain about it, I do not see a benefit of this comment.

florian-glombik commented on src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:
```diff
@@ -205,15 +204,19 @@ export class ProgrammingExerciseUpdateComponent implements AfterViewInit, OnDest
      * used in the template to display warnings.
      */
     refreshAuxiliaryRepositoryChecks() {
+        if (!this.programmingExercise?.auxiliaryRepositories) {
```
Fixed some errors here that were spamming the console due to undefined issues

florian-glombik commented on src/test/java/de/tum/in/www1/artemis/exercise/programming/ContinuousIntegrationTestService.java:
```diff
@@ -102,10 +104,6 @@ public ProgrammingExerciseStudentParticipation getParticipation() {
         return participation;
     }
 
-    public LocalRepository getLocalRepo() {
```
Method was unused

florian-glombik commented on src/main/webapp/i18n/en/programmingExercise.json:
```diff
@@ -436,15 +436,17 @@
             "preview": {
                 "label": "Preview",
-                "tooltip": "For every programming exercise, repositories and build plans will be generated. Here you can find their generated names based on the short name of the exercise (and course)",
+                "tooltip": "For every programming exercise, repositories and build plans will be generated. Here, you can find their generated names based on the short name of the exercise (and course)",
```
Added a comma after "Here", that's why there is no change in the german translation

florian-glombik commented on src/test/java/de/tum/in/www1/artemis/exercise/programming/ProgrammingExerciseLocalVCLocalCIIntegrationTest.java:
```diff
@@ -239,4 +240,20 @@ void testImportProgrammingExercise() throws Exception {
         localVCLocalCITestService.testLatestSubmission(solutionParticipation.getId(), null, 13, false);
     }
+
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "tutor1", roles = "TUTOR")
+    void testGetCheckoutDirectories() throws Exception {
+        RepositoriesCheckoutDirectoryDTO checkoutDirectoryDTO = request.get("/api/programming-exercises/repository-checkout-directories?programmingLanguage=JAVA", HttpStatus.OK,
+                RepositoriesCheckoutDirectoryDTO.class);
+        assertThat(checkoutDirectoryDTO.exerciseCheckoutDirectory()).isEqualTo("/assignment");
+        assertThat(checkoutDirectoryDTO.solutionCheckoutDirectory()).isEqualTo("/assignment");
+        assertThat(checkoutDirectoryDTO.testCheckoutDirectory()).isEqualTo("/");
+    }
```
`TUTOR` does not exist as role, it is called "TA"

florian-glombik commented on src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIService.java:
```diff
@@ -216,4 +221,21 @@ public boolean checkIfBuildPlanExists(String projectKey, String buildPlanId) {
         return name.toUpperCase().replaceAll("[^A-Z0-9]", "");
     }
+
+    @Override
+    public RepositoriesCheckoutDirectoryDTO getCheckoutDirectories(ProgrammingLanguage programmingLanguage) {
+        String exerciseCheckoutDirectory = ContinuousIntegrationService.RepositoryCheckoutPath.ASSIGNMENT.forProgrammingLanguage(programmingLanguage);
+        String solutionCheckoutDirectory = ContinuousIntegrationService.RepositoryCheckoutPath.SOLUTION.forProgrammingLanguage(programmingLanguage);
+        String testCheckoutDirectory = ContinuousIntegrationService.RepositoryCheckoutPath.TEST.forProgrammingLanguage(programmingLanguage);
+
+        exerciseCheckoutDirectory = startPathWithRootDirectory(exerciseCheckoutDirectory);
+        solutionCheckoutDirectory = startPathWithRootDirectory(solutionCheckoutDirectory);
+        testCheckoutDirectory = startPathWithRootDirectory(testCheckoutDirectory);
+
+        return new RepositoriesCheckoutDirectoryDTO(exerciseCheckoutDirectory, solutionCheckoutDirectory, testCheckoutDirectory);
+    }
```
That will be a bad request then, `programmingLanguage` is strongly typed.

```
{
    "type": "https://www.jhipster.tech/problem/problem-with-message",
    "title": "Bad Request",
    "status": 400,
    "detail": "Required request parameter 'programmingLanguage' for method parameter type ProgrammingLanguage is not present",
    "path": "/api/programming-exercises/repository-checkout-directories",
    "message": "error.http.400"
}
```

florian-glombik commented on src/main/webapp/app/exercises/programming/manage/repositories-checkout-directories-dto.ts:
```diff
@@ -0,0 +1,8 @@
+}
+
+// TODO is the file placed correctly here?
+// TODO is this how we define DTOs? read docs!
```
No, we seem to define DTOs in the entities folder in the client side

florian-glombik commented:
Regarding https://github.com/ls1intum/Artemis/pull/8597#issuecomment-2119239075

The values are not hard-coded, so they change if the programming language is changed. Below, you can see how it is displayed for Haskell and Ocaml - does that resolve the issue you had in mind @Strohgelaender?

#### Haskell Exercise
![image](https://github.com/ls1intum/Artemis/assets/63976129/dab2763f-7335-40cc-8d84-d4cd4b8dcba0)

#### Ocaml Exercise
![image](https://github.com/ls1intum/Artemis/assets/63976129/665abfda-d646-4ea2-af0a-fccc0731242b)

florian-glombik commented on src/main/java/de/tum/in/www1/artemis/security/Role.java:
```diff
@@ -5,7 +5,7 @@
 public enum Role {
 
-    // NOTE: we will soon rename "USER" to "STUDENT" in the database and add a new role "EDITOR"
+    // NOTE: we will soon rename "USER" to "STUDENT" in the database
```
The `EDITOR` role was added by now

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/web/rest/programming/ProgrammingExerciseResource.java:
```diff
@@ -878,4 +879,18 @@ public ResponseEntity<BuildLogStatisticsDTO> getBuildLogStatistics(@PathVariable
+     * @return a DTO containing the checkout directories for the exercise, solution, and test repository for the requested programming language.
+     */
+    @GetMapping("programming-exercises/repository-checkout-directories")
+    @EnforceAtLeastTutor
```
Do tutors need this information? Is that displayed in the exercise view to tutors as well?

florian-glombik replied:
I only had it enabled for editors at first and changed it to tutor since we do not really make a difference on the page, and it is no sensible data. But it probably makes the life for tutors easier if they only see what they should really need 👍

JohannesStoehr commented on src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:
```diff
@@ -171,76 +183,105 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
+                        if (profileInfo) {
+                            if (
+                                this.programmingExercise.projectKey &&
+                                this.programmingExercise.templateParticipation &&
+                                this.programmingExercise.templateParticipation.buildPlanId
```
```suggestion
                                this.programmingExercise.templateParticipation?.buildPlanId
```

JohannesStoehr commented on src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:
```diff
@@ -171,76 +183,105 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
+                            }
+                            if (
+                                this.programmingExercise.projectKey &&
+                                this.programmingExercise.solutionParticipation &&
+                                this.programmingExercise.solutionParticipation.buildPlanId
```
```suggestion
                                this.programmingExercise.solutionParticipation?.buildPlanId
```

JohannesStoehr commented on src/main/webapp/app/exercises/programming/manage/services/programming-exercise.service.ts:
```diff
@@ -673,4 +675,21 @@ export class ProgrammingExerciseService {
+            })
+            .pipe(
+                map((checkoutDirectories: RepositoriesCheckoutDirectoriesDTO) => {
+                    checkoutDirectories.exerciseCheckoutDirectory = addLeadingSlashIfNotPresent(checkoutDirectories.exerciseCheckoutDirectory);
+                    checkoutDirectories.solutionCheckoutDirectory = addLeadingSlashIfNotPresent(checkoutDirectories.solutionCheckoutDirectory);
+                    checkoutDirectories.testCheckoutDirectory = addLeadingSlashIfNotPresent(checkoutDirectories.testCheckoutDirectory);
```
Your server code already verifies that there is a slash. Do we need to do this here again?

JohannesStoehr commented on src/main/webapp/app/exercises/programming/manage/update/programming-exercise-plans-and-repositories-preview.component.ts:
```diff
@@ -1,20 +1,83 @@
+        }
+
+        if (
+            this.isLocal &&
+            this.programmingExerciseCreationConfig &&
+            this.programmingExerciseCreationConfig.selectedProgrammingLanguage &&
+            changes.programmingExerciseCreationConfig &&
+            changes.programmingExerciseCreationConfig.currentValue &&
+            changes.programmingExerciseCreationConfig.currentValue.selectedProgrammingLanguage !==
+                changes.programmingExerciseCreationConfig.previousValue?.selectedProgrammingLanguage
```
```suggestion
            this.isLocal &&
            this.programmingExerciseCreationConfig?.selectedProgrammingLanguage &&
            changes.programmingExerciseCreationConfig?.currentValue &&
            changes.programmingExerciseCreationConfig.currentValue.selectedProgrammingLanguage !==
                changes.programmingExerciseCreationConfig.previousValue?.selectedProgrammingLanguage
```

JohannesStoehr commented on src/main/webapp/i18n/en/programmingExercise.json:
```diff
@@ -436,15 +436,17 @@
                 "label": "Preview",
-                "tooltip": "For every programming exercise, repositories and build plans will be generated. Here you can find their generated names based on the short name of the exercise (and course)",
+                "tooltip": "For every programming exercise, repositories and build plans will be generated. Here, you can find their generated names based on the short name of the exercise (and course)",
+                "tooltipLocalVC": "Here, you can find the repositories linked to the exercise and their respective checkout directories during the Build Execution.",
```
```suggestion
                "tooltipLocalVC": "Here, you can find the repositories linked to the exercise and their respective checkout directories during the build execution.",
```

JohannesStoehr commented on src/main/webapp/i18n/en/programmingExercise.json:
```diff
@@ -436,15 +436,17 @@
                 "templateBuildPlanTooltip": "The build plan of the template. Runs all tests on the exercise template. Used as a reference in which all tests must fail",
-                "solutionBuildPlanTooltip": "The build plan of the solution. Runs all tests on the exercise solution. Used as a reference in which all tests must succeed"
+                "solutionBuildPlanTooltip": "The build plan of the solution. Runs all tests on the exercise solution. Used as a reference in which all tests must succeed",
+                "checkoutDirectoryDuringBuildExecution": "Checkout Directory during Build Execution"
```
Should this be in title case?

florian-glombik replied:
It as intentional but I do not have a strong opinion on the capitalization

florian-glombik commented on src/main/java/de/tum/in/www1/artemis/service/connectors/ci/ContinuousIntegrationService.java:
Thanks, this was not clear to me from the code. 

I have adjusted the error message and added some Javadoc, so this should be easier for future developers to understand.

florian-glombik commented on src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-repository-and-build-plan-details.component.ts:
```diff
@@ -0,0 +1,62 @@
+    checkoutDirectories?: CheckoutDirectoriesDto;
+    auxiliaryRepositoryCheckoutDirectories: string[] = [];
+
+    private updateCheckoutDirectories() {
+        if (!this.programmingLanguage) {
+            return;
+        }
+
+        this.programmingExerciseServiceSubscription = this.programmingExerciseService
+            .getCheckoutDirectoriesForProgrammingLanguage(this.programmingLanguage)
+            .subscribe((checkoutDirectories) => {
+                this.checkoutDirectories = checkoutDirectories;
+            });
+    }
```
No, these errors should be caught by the `alert.service.ts`

florian-glombik commented on src/main/webapp/app/detail-overview-list/detail-overview-list.component.html:
```diff
@@ -102,9 +102,7 @@ <h3 class="section-headline" [id]="headlinesRecord[section.headline]">{{ section
                                                 </a>
                                             }
                                             <div class="auxiliaryRepositoryDescription">
-                                                @if (auxiliaryRepository.checkoutDirectory) {
```
We do not need to display the checkout directory twice _(you can now retrieve this information from the checkout directory table)_

florian-glombik commented on src/main/webapp/app/exercises/programming/manage/update/programming-exercise-plans-and-repositories-preview.component.html:
```diff
@@ -1,67 +0,0 @@
-@if (programmingExercise && programmingExercise.shortName) {
```
This component is now under `shared/build-details` as we want to display it on the edit and details page.

It is also split into two components now:
`programming-exercise-build-plan-checkout-directories.component` and `programming-exercise-repository-and-build-plan-details.component`

pzdr7 reviewed (changes requested):
Code looks good! 👍 
I left a few comments

florian-glombik commented on src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-repository-and-build-plan-details.component.html:
```diff
@@ -0,0 +1,80 @@
+            <hr class="pt-0 pb-0 mt-1 mb-1" />
+            <ul>
+                <li>
+                    <code class="text-white badge bg-primary label-preview"> {{ courseShortName?.toUpperCase() }}{{ programmingExercise?.shortName?.toUpperCase() }} -BASE </code>
+                    <jhi-help-icon text="artemisApp.programmingExercise.preview.templateBuildPlanTooltip" />
+                </li>
+                <li>
+                    <code class="text-white badge bg-primary label-preview">
+                        {{ courseShortName?.toUpperCase() }}{{ programmingExercise?.shortName?.toUpperCase() }} -SOLUTION
```
Good catch, I did not look closely at the code that I just moved there!

florian-glombik commented on src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-build-plan-checkout-directories.component.html:
```diff
@@ -0,0 +1,24 @@
+@if (checkoutDirectories) {
+    <hr class="pt-0 pb-0 mt-1 mb-1" />
+    <ul>
+        <li [style.visibility]="checkoutDirectories.exerciseCheckoutDirectory ? '' : 'hidden'">
```
With hidden unfortunately the element in the DOM seems to be removed (in the second column the "/" should be at test directory and "assignment" at solution

![image](https://github.com/ls1intum/Artemis/assets/63976129/64ef49af-030c-4e6a-8aab-d67e5e50603b)

pzdr7 reviewed (approved):
Thanks for implementing the changes! Code LGTM

florian-glombik commented on src/test/javascript/spec/component/programming-exercise/programming-exercise-plans-and-repositories-preview.component.spec.ts:
```diff
@@ -0,0 +1,189 @@
+    });
+
+    it('should update auxiliary checkout repository directories', () => {
+        component.programmingExercise!.auxiliaryRepositories = [{ checkoutDirectory: 'assignment/sut' } as AuxiliaryRepository];
```
It is defined in @BeforeEach

florian-glombik commented on src/test/javascript/spec/component/programming-exercise/programming-exercise-plans-and-repositories-preview.component.spec.ts:
```diff
@@ -0,0 +1,189 @@
+
+        fixture.detectChanges();
+
+        const OCAML_SOLUTION_CHECKOUT_DIRECTORY_REGEX = RegExp('.*\\/assignment, *\\/solution.*');
```
I think that RegExp is better to read

florian-glombik commented on src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-repository-and-build-plan-details.component.ts:
```diff
@@ -0,0 +1,66 @@
+            this.checkoutDirectorySubscription.unsubscribe();
+        }
+
+        this.checkoutDirectorySubscription = this.programmingExerciseService
+            .getCheckoutDirectoriesForProgrammingLanguage(this.programmingLanguage)
+            .subscribe((checkoutDirectories) => {
+                this.checkoutDirectories = checkoutDirectories;
+            });
```
No.
`alert.service.ts` catches the errors

florian-glombik commented on src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-repository-and-build-plan-details.component.ts:
```diff
@@ -0,0 +1,66 @@
+import { Component, Input, OnChanges, OnDestroy, OnInit, SimpleChanges } from '@angular/core';
+import { getCourseFromExercise } from 'app/entities/exercise.model';
+import type { ProgrammingExercise, ProgrammingLanguage } from 'app/entities/programming-exercise.model';
+import { ProgrammingExerciseService } from 'app/exercises/programming/manage/services/programming-exercise.service';
+import { Subscription } from 'rxjs';
+import type { CheckoutDirectoriesDto } from 'app/entities/checkout-directories-dto';
```
The other imports dont work with `type`

az108 commented on src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:
```diff
@@ -375,6 +408,15 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
                             showOpenLink: !this.localVCEnabled,
                         },
                     },
+                this.localCIEnabled && {
```
i think that this should only be defined for editors+

az108 reviewed (approved):
Works as expected, tested in testing session together and tested it myself aswell and did codereview together with florian

florian-glombik commented on src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-repository-and-build-plan-details.component.ts:
```diff
@@ -0,0 +1,66 @@
+        this.courseShortName = getCourseFromExercise(this.programmingExercise)?.shortName;
+    }
+
+    ngOnInit() {
```
I moved the lifecycle methods to the top

florian-glombik commented on src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-repository-and-build-plan-details.component.ts:
```diff
@@ -0,0 +1,68 @@
+            return;
+        }
+
+        this.checkoutDirectorySubscription?.unsubscribe(); // might be defined from previous method execution
+
+        const CHECKOUT_SOLUTION_REPOSITORY_DEFAULT = true;
+        this.checkoutDirectorySubscription = this.programmingExerciseService
+            .getCheckoutDirectoriesForProgrammingLanguage(this.programmingLanguage, this.checkoutSolutionRepository ?? CHECKOUT_SOLUTION_REPOSITORY_DEFAULT)
+            .subscribe((checkoutDirectories) => {
+                this.checkoutDirectories = checkoutDirectories;
+            });
```
Still caught by the alert service

florian-glombik commented:
Changes since last reviews: 
* verified checkout directories, noticed that solution will not be checked out in two repositories for the solution build plan => adjusted the DTO accordingly (no string array anymore)
* endpoint now considers if the solution checkout directory shall be checked out or not (for Haskell and OCaml)

pzdr7 commented on src/test/javascript/spec/component/programming-exercise/programming-exercise-plans-and-repositories-preview.component.spec.ts:
```diff
@@ -0,0 +1,173 @@
+    };
+
+    beforeEach(() => {
+        TestBed.configureTestingModule({
+            declarations: [ProgrammingExerciseRepositoryAndBuildPlanDetailsComponent, MockComponent(HelpIconComponent), ProgrammingExerciseBuildPlanCheckoutDirectoriesComponent],
+            providers: [{ provide: ProgrammingExerciseService, useClass: MockProgrammingExerciseService }],
+        })
```
The tests fail because the component fixture cannot be created. Here's an implementation that worked for me locally (please double-check it).

```ts
        TestBed.configureTestingModule({
            declarations: [ProgrammingExerciseRepositoryAndBuildPlanDetailsComponent, MockComponent(HelpIconComponent), ProgrammingExerciseBuildPlanCheckoutDirectoriesComponent],
            imports: [ArtemisSharedComponentModule, ArtemisSharedCommonModule],
            providers: [
                { provide: ProgrammingExerciseService, useClass: MockProgrammingExerciseService },
                { provide: TranslateService, useClass: MockTranslateService },
            ],
        })
```

florian-glombik replied:
Sorry for not waiting until the tests have passed - I should have double checked them after the conversion of the module to a standalone component!

Thanks for even investigating the needed import and providers 🙏

pzdr7 commented on src/test/javascript/spec/component/programming-exercise/programming-exercise-plans-and-repositories-preview.component.spec.ts:
```diff
@@ -0,0 +1,173 @@
+                component.programmingExercise = { id: 1, shortName: 'shortName' } as ProgrammingExercise;
+                component.isLocal = true;
+
+                jest.spyOn(programmingExerciseService, 'getCheckoutDirectoriesForProgrammingLanguage').mockImplementation((programmingLanguage: ProgrammingLanguage) => {
```
Do you think it could make sense to include `checkoutSolution: boolean` here too since the component has changed to include this new detail? I suppose you'd have to add another test in that case

pzdr7 commented on src/test/javascript/spec/helpers/mocks/service/mock-programming-exercise.service.ts:
```diff
@@ -26,4 +27,5 @@ export class MockProgrammingExerciseService {
     generateStructureOracle = (exerciseId: number) => of({});
     unlockAllRepositories = (exerciseId: number) => of({});
     getDiffReportForCommits = (exerciseId: number, participationId: number, olderCommitHash: string, newerCommitHash: string, repositoryType: string) => of({});
+    getCheckoutDirectoriesForProgrammingLanguage = (programmingLanguage: ProgrammingLanguage) => of();
```
The signature has changed:

```suggestion
    getCheckoutDirectoriesForProgrammingLanguage = (programmingLanguage: ProgrammingLanguage, checkoutSolution: boolean) => of();
```

pzdr7 reviewed (changes requested):
The code looks good overall. I left a few comments on things I've noticed.

Also, the test setup is currently not instantiating the ProgrammingExerciseRepositoryAndBuildPlanDetailsComponent correctly (see below for one version that worked for me locally). Can you take a look?

florian-glombik commented on src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-repository-and-build-plan-details.component.ts:
```diff
@@ -0,0 +1,73 @@
+
+    courseShortName?: string;
+    checkoutDirectories?: CheckoutDirectoriesDto;
+    auxiliaryRepositoryCheckoutDirectories: string[] = [];
```
Good catch, leftover after refactoring - the variable `auxiliaryRepositoryCheckoutDirectories` is unused

florian-glombik commented:
Sorry for the confusing diff in the `.spec.ts` files, I already contacted @undernagruzez as I think these findings are false positives.

az108 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java:
```diff
@@ -252,4 +254,9 @@ public void createProjectForExercise(ProgrammingExercise programmingExercise) th
+
+    @Override
+    public CheckoutDirectoriesDTO getCheckoutDirectories(ProgrammingLanguage programmingLanguage, boolean checkoutSolution) {
+        throw new UnsupportedOperationException("Method not implemented, consult the build plans in Jenkins for the checkout directories.");
```
Wouldnt it be better to say. "Method not implemented, consult the build plans in Jenkins to receive more information on the checkout directories."

florian-glombik replied:
I adjusted it to "Method not implemented, consult the build plans in GitLab for more information on the checkout directories." - I think this fixes the issue?

az108 replied:
yes thanks :)

az108 reviewed (approved):
Changes adressed, LGTM

pzdr7 reviewed (approved):
Thanks for addressing my feedback. Code LGTM 👍

sarpsahinalp reviewed (approved):
Tested locally works as expected

BBesrour reviewed (approved):
Tested on TS2, works as described.
